drakrun/drakrun/regression.py

```diff
@@ -0,0 +1,204 @@
+        super().__init__(config)
+
+    def analyze_dumps(self, sample, dump_dir):
+        manager = ExtractManager(ExtractorModules(self.config.config['draktestd']['modules']))
+        dumps = Path(dump_dir) / "dumps"
+        family = None
+        for f in dumps.glob("*.metadata"):
+            with open(f, "rb") as metafile:
+                metadata = json.load(metafile)
+            va = int(metadata["DumpAddress"], 16)
+            name = dumps / metadata["DataFileName"]
+
+            with changedLogLevel(logging.getLogger(), logging.ERROR):
+                res = manager.push_file(name, base=va)
+                family = family or res
+        return family
+
+    def process(self, task: Task):
+        dumps = task.get_resource("dumps.zip")
+        sample = task.get_resource("sample")
+        with dumps.extract_temporary() as temp:
+            family = self.analyze_dumps(sample, temp)
+
+            testcase = TestCase.from_json(task.payload["testcase"])
+            expected_family = testcase.ripped
+
+            if family is None or expected_family != family:
+                self.log.error(f"Failed to rip {sample.sha256}. Expected {expected_family}, ripped {family}")
+                result = 'FAIL'
+            else:
+                self.log.info(f"Ripping {sample.sha256} OK: {family}")
+                result = 'OK'
+
+            out_res = json.dumps({"sample": sample.sha256, "family": {"expected": expected_family, "ripped": family}, "result": result})
+
+            task = Task({"type": "analysis-test-result", "kind": "drakrun"})
+            res = LocalResource(name=self.current_task.root_uid, bucket='draktestd', content=out_res)
+            res._uid = res.name
+            task.add_payload("result", res)
+            self.send_task(task)
+
+    @classmethod
+    def args_parser(cls):
+        parser = super().args_parser()
+        parser.add_argument("tests")
+        parser.add_argument("--timeout", type=int)
+        return parser
+
+    @classmethod
+    def main(cls):
+        conf_path = os.path.join(ETC_DIR, "config.ini")
+        config = patch_config(Config(conf_path))
+
+        consumer = RegressionTester(config)
+
+        if not consumer.backend.minio.bucket_exists('draktestd'):
+            consumer.backend.minio.make_bucket(bucket_name='draktestd')
+
+        consumer.loop()
+
+    @classmethod
+    def get_finished_tasks(cls, rs, root_uids):
+        running = {}
+
+        for root_uid in root_uids:
+            running[root_uid] = 0
+
+        for task in rs.keys('karton.task:*'):
+            obj_s = rs.get(task)
+
+            if not obj_s:
+                continue
+
+            obj = json.loads(obj_s)
+
+            if obj['root_uid'] not in running:
+                continue
+
+            if obj['status'] not in ['Finished', 'Crashed']:
+                running[obj['root_uid']] = running[obj['root_uid']] + 1
+
+        return [root_uid for root_uid, cnt in running.items() if cnt == 0]
```

Comment:
I think that is could be simplified by using sets instead of a dictionary.
That is, the output of this function is a set difference between root_uids and tasks that are still being processed.